src/storage.rs

```diff
@@ -0,0 +1,195 @@
+use std::ops::{Deref, DerefMut};
+use std::fmt;
+
+/// TODO: Write some docs
+#[derive(Debug)]
+pub struct StorageReadGuard<'a, T: Serialize + DeserializeOwned + 'a> {
+    cache: &'a Storage<T>,
+}
+
+impl<'a, T: Serialize + DeserializeOwned> StorageReadGuard<'a, T> {
+    fn new(cache: &'a Storage<T>) -> Self {
+        Self { cache }
+    }
+}
+
+impl<'a, T: Serialize + DeserializeOwned + 'a> Deref for StorageReadGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.cache.data
+    }
+}
+
+impl<'a, T: 'a + Serialize + DeserializeOwned + fmt::Display> fmt::Display for StorageReadGuard<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+
+/// TODO: Write some docs
+#[derive(Debug)]
+pub struct StorageWriteGuard<'a, T: Serialize + DeserializeOwned + 'a> {
+    cache: &'a mut Storage<T>,
+}
+
+impl<'a, T: Serialize + DeserializeOwned> StorageWriteGuard<'a, T> {
+    fn new(cache: &'a mut Storage<T>) -> Self {
+        Self { cache }
+    }
+}
+
+impl<'a, T: Serialize + DeserializeOwned> Drop for StorageWriteGuard<'a, T> {
+    fn drop(&mut self) {
+        self.cache.file.seek(SeekFrom::Start(0)).unwrap();
+        self.cache.file.set_len(0).unwrap();
+        self.cache.file.write(to_string(&self.cache.data).unwrap().as_bytes()).unwrap();
+    }
+}
+
+impl<'a, T: Serialize + DeserializeOwned + 'a> Deref for StorageWriteGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.cache.data
+    }
+}
+
+impl<'a, T: Serialize + DeserializeOwned + 'a> DerefMut for StorageWriteGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.cache.data
+    }
+}
+
+impl<'a, T: 'a + Serialize + DeserializeOwned + fmt::Display> fmt::Display for StorageWriteGuard<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        (**self).fmt(f)
+    }
+}
+
+/// TODO: Write some docs
+#[derive(Debug)]
+pub struct Storage<T: Serialize + DeserializeOwned> {
+    data: T,
+    file: File,
+}
+
+impl<T: Serialize + DeserializeOwned> Storage<T> {
+    pub fn new_with_default<P: Into<String>, F: Fn() -> T>(
+        path: P,
+        default: F,
+    ) -> Result<Self, String> {
+        let path = path.into();
+
+        // Read the file first, to see if there's any existing data
+        let data = match File::open(path.clone()) {
+            Ok(mut f) => {
+                let mut contents = String::new();
+
+                f.read_to_string(&mut contents)
+                    .map_err(|err| format!("Couldn't read file: {}", err))?;
+
+                from_str(&contents)
+                    .map_err(|err| format!("Couldn't read file: {}", err))?
+            }
+            Err(_) => default()
+        };
+
+        // Then open the file again and truncate, preparing it to be written to
+        Ok(Self { data, file: OpenOptions::new()
+            .read(true)
+            .write(true)
+            .truncate(true)
```

Comment:
I do not think truncating the file immediately is a good idea. This would mean that if the process were to crash after reading the file into memory, but before a new value is written, the persisted data would be lost.